### PR TITLE
Add dexFromCAtom to the C API

### PR DESCRIPTION
**Pull request by apaszke**
This makes it possible to use `CAtom` as a common language for getting
values both into and out of Dex.

cc @oxinabox

### CODE REVIEW

File: python/dex/__init__.py [hunk 1]
```diff
@@ -52,15 +52,35 @@ def eval(expr: str, module=prelude, _env=None):
...
+  def __init__(self, value):
+    catom = api.CAtom()
+    if isinstance(value, int):
+      catom.tag = 0
+      catom.value.tag = 1
+      catom.value.value = ctypes.c_int(value)
+    elif isinstance(value, float):
+      catom.tag = 0
+      catom.value.tag = 4
+      catom.value.value = ctypes.c_float(value)
```
Comment (oxinabox) [inline python/dex/__init__.py:70]: Why only `Int64` and `Float32`?

I thought floats in python were `Float64`s precision.
And `Int`'s were arbitrary precision?

I have no idea how ctypes handles these though

File: python/dex/__init__.py [hunk 2]
```diff
@@ -52,15 +52,35 @@ def eval(expr: str, module=prelude, _env=None):
...
+    else:
+      raise ValueError("Can't convert given value to a Dex Atom")
+    self.module = prelude
```
Comment (oxinabox) [inline python/dex/__init__.py:73]: Shouldn't this be an optional argument, that is defaulted to `prelude` ?
Or is it fine because as a literal it is harmless to add it to the prelude and every other module is sure to have access to things from `prelude` ?

File: python/dex/api.py [hunk 3]
```diff
@@ -17,11 +17,22 @@ def tagged_union(name: str, members: List[type]):
...
 
-CLit = tagged_union("Lit", [ctypes.c_int64, ctypes.c_int32, ctypes.c_int8, ctypes.c_double, ctypes.c_float])
+CLit = tagged_union("Lit", [
+  ctypes.c_int64,
+  ctypes.c_int32,
+  ctypes.c_uint8,
+  ctypes.c_double,
+  ctypes.c_float,
+  ctypes.c_uint32,
+  ctypes.c_uint64
```
Comment (oxinabox) [inline python/dex/api.py:34]: A matching change to this seems like it would be needed in the Julia code.
Which would be to add `UInt32` and `UInt64` to the ends of the list of types here:
https://github.com/google-research/dex-lang/blob/07550931f59bc5b4148e47ec940a177e26518872/julia/src/api_types.jl#L43

File: python/tests/jax_test.py [hunk 4]
```diff
@@ -72,12 +72,12 @@ def test_vmap_unbatched_array(self):
...
     add_two = primitive(dex.eval(r'\x:((Fin 2)=>Float). for i. x.i + 2.0'))
-    x = jnp.linspace([0, 3], [5, 8], num=4, dtype=jnp.float32)
+    x = jnp.linspace(jnp.array([0, 3]), jnp.array([5, 8]), num=4, dtype=jnp.float32)
     np.testing.assert_allclose(jax.jit(jax.vmap(add_two))(x), x + 2.0)
 
   def test_vmap_jit_nonzero_index(self):
     add_two = primitive(dex.eval(r'\x:((Fin 4)=>Float). for i. x.i + 2.0'))
-    x = jnp.linspace([0, 3], [5, 8], num=4, dtype=jnp.float32)
+    x = jnp.linspace(jnp.array([0, 3]), jnp.array([5, 8]), num=4, dtype=jnp.float32)
     np.testing.assert_allclose(
```
Comment (oxinabox) [inline python/tests/jax_test.py:81]: All changes in this file are unrelated right?

File: python/dex/__init__.py (hunk 1 above)
Comment (apaszke) [reply python/dex/__init__.py:70]: Just because I wrote this out quickly to unblock you and this was good enough for a simple test. I decided to downcast `int` and `float`, because `Int` and `Float` in Dex are 32-bit. If someone's serious about precision they can always use NumPy scalars to bundle the dtype they want to get on the Dex side.

File: python/dex/__init__.py (hunk 2 above)
Comment (apaszke) [reply python/dex/__init__.py:73]: Yeah the `module` should be a set of bindings that includes all definitions of free variables appearing in the atom. But if the atom is a literal then it has no free variables, so any module would do. But a prelude seems like a reasonable default.

File: python/dex/api.py (hunk 3 above)
Comment (apaszke) [reply python/dex/api.py:34]: Well it's not even a bug fix. The old code was correct, it just couldn't handle some of the cases.

File: python/tests/jax_test.py (hunk 4 above)
Comment (apaszke) [reply python/tests/jax_test.py:81]: Yes, they're due to JAX making the NumPy API stricter than it used to be.

**Review by oxinabox — APPROVED** [review]
